src/test/java/teammates/test/driver/EmailAccount.java

```diff
@@ -26,37 +34,117 @@ private EmailAccount() {
     }
 
     /**
-     * Retrieve registration key sent to Gmail inbox. After retrieving, marks
-     * the email as read.
+     * Retrieve registration key sent to Gmail inbox. After retrieving, marks the email as read.
      *
      * @return registration key (null if cannot be found).
      */
-    public static String getRegistrationKeyFromGmail(String username, String password, String courseName, String courseId)
+    public static String getRegistrationKeyFromGmail(String username, String courseName, String courseId)
             throws IOException, MessagingException {
-        Folder inbox = getGmailInbox(username, password);
-        Message[] messages = getMessages(inbox);
-
-        // Loop over up to 5 unread messages at the top of the inbox
-        int maxEmailsToCheck = Math.min(messages.length, 5);
-        for (int i = messages.length - 1; i >= messages.length - maxEmailsToCheck; i--) {
-            Message message = messages[i];
-
-            if (isRegistrationEmail(message, courseName, courseId)) {
-                String body = getEmailBody(message);
-                message.setFlag(Flags.Flag.SEEN, true);
-                inbox.close(true);
-                return getKey(body);
+
+        Gmail service = new GmailServiceMaker(username).makeGmailService();
+
+        ListMessagesResponse listMessagesResponse;
+        while (true) {
+            try {
+                // Get last 5 emails received by the user as there may be other emails received. However, this may fail
+                // unexpectedly if there are 5 additional emails received excluding the one from TEAMMATES.
+                listMessagesResponse = service.users().messages().list(username).setMaxResults(5L)
+                        .setQ("is:UNREAD").execute();
+                break;
+            } catch (GoogleJsonResponseException e) {
+                if (e.getDetails().getCode() == HttpStatusCodes.STATUS_CODE_FORBIDDEN) {
+                    System.out.println(e.getDetails().getMessage());
+                    service = new GmailServiceMaker(username, true).makeGmailService();
+                } else {
+                    throw e;
+                }
+            }
+        }
+
+        final List<Message> messageStubs = listMessagesResponse.getMessages();
+        if (messageStubs != null) {
+            for (Message messageStub : messageStubs) {
+                final Message message = service.users().messages().get(username, messageStub.getId()).setFormat("raw")
+                        .execute();
+
+                final MimeMessage email = convertFromMessageToMimeMessage(message);
+
+                if (isStudentCourseJoinRegistrationEmail(email, courseName, courseId)) {
```

Comment:
I am not sure how to put it but I don't think the normal path should not be nested at all is a rule.

For this case, we are searching for something in a loop, and searching for things in a loop is a common scenario.

Otherwise, we would have to invert the logic for almost every loop we write, use guard clauses, and then go to the actual logic itself. I do not think that helps in readability. In the link you gave itself, the flow is purportedly shown to be very convoluted, not to mention it is not in a loop.

So maybe looking at this code itself, I interpreted it this way:
```
Search for student course join registration emails and get key
```
On the other hand, using the guard clause has the following meaning for me:
```
Search for non course join registration emails and skip them.
OTHERWISE
Get the key from the emails
```

Unless there are clear benefits as in the example you have shown in the link, I do not think strictly adhering to using the guard clause in every case is right.

I would even argue that using guard clause here decreases readability because I have to remember the guard clause, and then remember that in the scenario where it is inverted, I will get the key from email.

Whereas a case I might use a guard clause might be `if (isMalformed(email)) continue;`
In that case I do not have to remember that the email is malformed because it is not normal, and hence I don’t have to remember the guard clause.

So I think in order to use a guard clause maybe a condition should be:
> If it is an exceptional flow that need not be considered anymore because it is completely handled